proposed/simple-csharp-programs-motivation-walkthrough.md

```diff
@@ -0,0 +1,311 @@
+exports.log = {
+  dim: (text) => console.log(chalk.dim(text)),
+  success: (text) => console.log(chalk.green(text))
+}
+
+// Generates a fileName with current time. e.g. fake-names-19:29:17.json
+exports.getFileName = () => `fake-names-${currentTime}.json`
+```
+
+**program.js**
+
+```js
+const { log, getFileName } = require('../lib/utils')
+
+// use stuff from `utils`
+```
+
+Similarly, in F# there is the `#load "path-to-file.fsx"` directive.
+
+This leads to some questions:
+
+* Are multiple file supported? That is, can I define a `server.cs` file that is my web service, and a `data.cs` that handles some data manipulation that is called from code inside of `server.cs`?
+* If multiple files are supported, what are the guardrails? How can people know if they made a mistake and fix it?
+
+### Coding against a framework alternative syntax
+
+The walkthrough proposes the shebang `#!/usr/bin/share/dotnet-aspnet`, but the .NET CLI doesn't have a command to "pull in the ASP.NET Core references", so it wouldn't map to something like that.
+
+An alternative could be to extend `#r` like so:
+
+* `#r "framework:aspnet"`
+* `#r "sdk:aspnet"`
+* etc.
+
+This could cleanly separate the "tell the compiler which references to pull in" world from the "short-cut having to use `dotnet`" world like so:
+
+```csharp
+#!/usr/bin/share/dotnet
+#r "framework:aspnet"
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+### Shebangs and their semantics
+
+The shebangs in the walkthrough all point at `/usr/bin/share/dotnet`. Substitute a path for your appropriate installation of .NET on your operating system of choice.
+
+What this means is that it identifies the .NET environment and implies a default set of references available in the file. However, it does not bring in other things like ASP.NET or other frameworkreferences. It is a contextual, "this is how I run" sort of directive.
+
+It could be extended/changed in a few ways:
+
+* `/usr/bin/share/dotnet run` makes it explicit that the command is to run the file
+* `/usr/bin/share/dotnet-aspnet` is an alternative to `#!aspnet` that brings ASP.NET Core into scope for the file (substitute a different framework/sdk here)
+
+In effect, the shebang's semantics could span a spectrum of "I need the path and command to literally execute this file" towards "I add context that affects the references available to this file".
+
+We'll need to decide if they can affect the set of references available in the file or not.
+
+### Build output
+
+The walkthrough states that the build output is a single executable file in the current working directory. This is based on the hypothesis that this is the least surprising place to put the executable.
+
+* Is this hypothesis based on reality?
+* Is this still true if you pull in a package?
+* Is this still true if you pull in a framework? (e.g., ASP.NET Core is hundreds of assemblies, where do they go?)
```

Comment:
Most of those assemblies live in the shared framework, they do not end up in your bin dir by default.